backend/azure/dtmb/dtmb.go

```diff
@@ -0,0 +1,531 @@
+
+	var historyRequest dtmbprotos.GetOrchestrationHistoryRequest
+	if cachedEvents == nil {
+		historyRequest = dtmbprotos.GetOrchestrationHistoryRequest{
+			OrchestrationId: orchestrationID,
+		}
+	} else {
+		historyRequest = dtmbprotos.GetOrchestrationHistoryRequest{
+			OrchestrationId:        orchestrationID,
+			LastItemSequenceNumber: cachedEvents[len(cachedEvents)-1].GetSequenceNumber(),
+			LastItemEventHash:      cachedEvents[len(cachedEvents)-1].GetEventHash(),
+		}
+	}
+
+	res, err := d.workerClient.GetOrchestrationHistory(
+		ctx,
+		&historyRequest,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	events := res.GetEvent()
+
+	// cache the newly received events
+	d.orchestrationHistoryCache.AddHistoryEventsForOrchestrationID(orchestrationID, events)
+
+	if cachedEvents != nil {
+		// merge the sever events with the cached events
+		events = append(cachedEvents, events...)
+	}
+	return events, err
+}
+
+// GetOrchestrationWorkItem gets a pending work item from the task hub or returns [ErrNoWorkItems]
+// if there are no pending work items.
+func (d *dtmb) GetOrchestrationWorkItem(ctx context.Context) (*backend.OrchestrationWorkItem, error) {
+	var ret *backend.OrchestrationWorkItem = nil
+	item := d.orchestrationQueue.Dequeue()
+	if item == nil {
+		return nil, backend.ErrNoWorkItems
+	}
+
+	cachedEvents, err := d.getOrchestrationHistory(ctx, item.OrchestrationId)
+	if err != nil {
+		return nil, err
+	}
+
+	// combine cached events with new events yet to be executed
+	allEvents := append(cachedEvents, item.GetNewEvents()...)
+
+	convertedEvents, err := utils.ConvertEvents(allEvents)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(convertedEvents) == 0 {
+		return nil, backend.ErrNoWorkItems
+	}
+
+	ret = &backend.OrchestrationWorkItem{
+		InstanceID: api.InstanceID(item.GetOrchestrationId()),
+		NewEvents:  convertedEvents,
+		RetryCount: int32(item.GetRetryCount()),
+		// RetryCount: TODO: Alessandro to implement
+		Properties: map[string]interface{}{
+			"CompletionToken":   item.GetCompletionToken(),
+			"OrchestrationName": item.GetName(),
+			"Version":           item.GetVersion(),
+			"ExecutionId":       item.GetExecutionId(),
+		},
+	}
+	return ret, nil
+}
+
+// GetOrchestrationRuntimeState gets the runtime state of an orchestration instance.
+func (d *dtmb) GetOrchestrationRuntimeState(ctx context.Context, workitem *backend.OrchestrationWorkItem) (*backend.OrchestrationRuntimeState, error) {
+	events, err := d.getOrchestrationHistory(ctx, string(workitem.InstanceID))
+	if err != nil {
+		return nil, err
+	}
+	historyEvents, err := utils.ConvertEvents(events)
+	if err != nil {
+		return nil, err
+	}
+
+	state := backend.NewOrchestrationRuntimeState(workitem.InstanceID, historyEvents)
+	return state, nil
+}
+
+// GetOrchestrationMetadata gets the metadata associated with the given orchestration instance ID.
+//
+// Returns [api.ErrInstanceNotFound] if the orchestration instance doesn't exist.
```

Comment:
I didn't see this error value being used, so just curious: how are we handling the case where the orchestration doesn't exist? Is there some value is `resp` that would indicate this?

Reply:
It's not totally clear what errors are returned in these cases. In general, returned errors from the API are something we will have to cover in a second pass. I haven't seen any details on what errors look like.

Reply:
If the orchestration doesn't exist the gRPC request will fail with a gRPC error (`err` will contain a gRPC error). You can get the status code from that error and it will be a "NotFound". More details are in the error description.

Reply:
In general, errors returned by the gRPC APIs follow the typical pattern for gRPC errors. For Go, this is the key method: https://pkg.go.dev/google.golang.org/grpc/status#Code